crates/ruff/src/settings/line_width.rs

```diff
@@ -0,0 +1,209 @@
+use ruff_cache::CacheKey;
+use serde::{Deserialize, Serialize};
+use unicode_width::UnicodeWidthChar;
+
+use ruff_macros::CacheKey;
+
+pub trait LineWidthState: Clone + Default + CacheKey {}
+
+#[derive(Clone, Copy, Debug, CacheKey)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+pub struct Length;
+
+impl Default for Length {
+    fn default() -> Self {
+        Self
+    }
+}
+
+#[derive(Clone, Copy, Default, CacheKey)]
+pub struct TabInfos {
+    column: usize,
+    tab_size: TabSize,
+}
+
+impl LineWidthState for Length {}
+impl LineWidthState for TabInfos {}
+
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, CacheKey)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[serde(from = "usize", into = "usize")]
+pub struct LineWidth<S: LineWidthState> {
```

Comment:
Use the Typestate Pattern to only have `new`, `add_str`, `add_char`, `add_width` on `Width` and not `LineLength`.

Reply:
http://cliffle.com/blog/rust-typestate/

Reply:
Do I understand it correctly that the main benefit of applying this pattern is that we can share the `width()` implementation between `Width` and `LineLength`. 

If that's the case, I would prefer to duplicate the method in both structs instead. It makes the code significantly easier to reason about and the duplication is minimal (results in the same generated code). 

It would also allow us to split the logic into multiple files. That would be beneficial because `Width` (as far as I understand) isn't used in any setting context. We should move it out of the `settings` module because of that.